File: Api/src/main/java/net/winternetwork/bedwars/api/score/Score.java
package net.winternetwork.bedwars.api.score;

import lombok.AllArgsConstructor;
import lombok.Getter;
import lombok.RequiredArgsConstructor;
import org.bukkit.Bukkit;
import org.bukkit.ChatColor;
import org.bukkit.entity.Player;
import org.bukkit.scoreboard.DisplaySlot;
import org.bukkit.scoreboard.Objective;
import org.bukkit.scoreboard.Scoreboard;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.function.Function;

@RequiredArgsConstructor
public class Score {

    private final String title;

    private final List<TableLine> lines = Collections.synchronizedList(new ArrayList<>());

    public void show(Player player) {
        Scoreboard score = player.getScoreboard();

        if (score == null)
            score = Bukkit.getScoreboardManager().getNewScoreboard();

        Objective obj = getObjective(player);

        obj.setDisplayName(c(title));
        obj.setDisplaySlot(DisplaySlot.SIDEBAR);

        updateLines(player);
        player.setScoreboard(score);
    }

    public void setLine(int line, String text) {
        lines.add(new TableLine(line, text, null));
    }

    private void updateLines(Player player) {
        Objective objective = getObjective(player);

        for (int i = lines.size(); i > 0; i--) {
            TableLine line = lines.get(i - 1);
            String text = c(line.getText());

            String score = scoreEntry(objective, line.getLine());
            if (score != null) {
                if (score.equalsIgnoreCase(text)) continue;
                objective.getScoreboard().resetScores(score);
            }
            objective.getScore(text).setScore(line.getLine());
        }
    }

    private String scoreEntry(Objective o, int score) {
        return o.getScoreboard()
                .getEntries()
                .stream()
                .filter(it -> o.getScore(it).getScore() == score)
                .findFirst()
                .orElse(null);
    }

    private Objective getObjective(Player player) {
        Scoreboard score = player.getScoreboard();

        Objective objective;

        if ((objective = score.getObjective(player.getName())) == null) {
            objective = score.registerNewObjective(player.getName(), "dummy");
        }

        return objective;
    }

    private String c(String c) {
        return ChatColor.translateAlternateColorCodes('&', c);
    }

    @Getter
    @AllArgsConstructor
    private static class TableLine {

        private final int line;
        private final String text;
        private final Function<Player, String> function;
    }
}

File: Api/src/main/kotlin/net/winternetwork/bedwars/api/score/Score.kt
package net.winternetwork.bedwars.api.score

import net.winternetwork.bedwars.api.util.color
import org.bukkit.Bukkit
import org.bukkit.entity.Player
import org.bukkit.scoreboard.DisplaySlot
import org.bukkit.scoreboard.Objective
import java.util.*

class Score(private val title: String) {

    private val lines = Collections.synchronizedList(ArrayList<TableLine>())

    fun show(player: Player) {
        val score = player.scoreboard ?: Bukkit.getScoreboardManager().newScoreboard

        getObjective(player).run {
            displayName = title.color
            displaySlot = DisplaySlot.SIDEBAR
        }

        updateLines(player)
        player.scoreboard = score

    }

    fun setLine(line: Int, text: String) {
        lines.add(TableLine(line, text))
    }

    private fun updateLines(player: Player) {
        val obj = getObjective(player)

        for (i in lines.size..0) {
            val line = lines[i]
            val text = line.text.color

            val score = scoreEntry(obj, line.line)
            if (score != null) {
                if (score.equals(text, true)) continue
                obj.scoreboard.resetScores(score)
            }

            obj.getScore(text).score = line.line
        }
    }

    private fun scoreEntry(obj: Objective, score: Int) =
            obj.scoreboard.entries.firstOrNull { obj.getScore(it).score == score }

    private fun getObjective(player: Player): Objective {
        val score = player.scoreboard

        return score.getObjective(player.name)
                ?: score.registerNewObjective(player.name, "dummy")
    }
}

internal data class TableLine(
        val line: Int,
        val text: String
)